dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java

```diff
@@ -1055,27 +1055,88 @@ public String getRepositoryId(Entry entry) {
         }
     }
 
+
+
     @Override
     public String getMainBranch(Entry entry, String repositoryId) {
         String mainBranch = null;
 
         // Get repository based on username and repo id
+        mainBranch = getDefaultBranch(repositoryId);
+        if (mainBranch == null) {
+            return null;
+        }
+        // Determine which branch to use for tool info
+        if (entry.getDefaultVersion() != null) {
+            mainBranch = getBranchNameFromDefaultVersion(entry);
+        }
+
+        return mainBranch;
+    }
+
+    @Override
+    public String getDefaultBranch(String repositoryId) {
         if (repositoryId != null) {
             try {
                 GHRepository repository = github.getRepository(repositoryId);
-                // Determine the default branch on Github
-                mainBranch = repository.getDefaultBranch();
+                // Determine the default branch on GitHub
+                return repository.getDefaultBranch();
             } catch (IOException e) {
                 LOG.error("Unable to retrieve default branch for repository " + repositoryId, e);
                 return null;
             }
         }
-        // Determine which branch to use for tool info
-        if (entry.getDefaultVersion() != null) {
-            mainBranch = getBranchNameFromDefaultVersion(entry);
-        }
+        return null;
+    }
 
-        return mainBranch;
+    /**
+     * Detect branches that include a .dockstore.yml using a variety of heuristics.
+     * <p>
+     * Heuristics include looking at the default branch, commits that just touched a .dockstore.yml at the end of a branch, etc. while trying to avoid too many
+     * API calls, particularly an unbounded number of calls based on factors that can change between repos (number of tags, branches, etc.)
```

Comment:
I was going to suggest you could also look at the most recent release, but it's unlikely that would have a .dockstore.yml, so probably not after all.